### PR TITLE
Change L1T DQM configuration to have stage1 QTest and configuration for stage1 heavy ion 

### PR DESCRIPTION
**Pull request by jingyucms**
change L1 DQM to have stage1 quality test and make stage1 heavy ion run configuration

### CODE REVIEW

**Comment by deguio** [comment]
please test

**Comment by deguio** [comment]
this PR looks OK and its 74 version has been used since a while in the online.
wait to see #11756 and #11923 rebased before approving

**Comment by deguio** [comment]
please test

**Comment by deguio** [comment]
please test

**Comment by deguio** [comment]
+1
this is final @davidlange6

**Comment by mulhearn** [comment]
+1

**Comment by slava77** [comment]
@cmsbuild please test
[comparisons never made it]

**Comment by slava77** [comment]
@deguio @mulhearn 
how essential is this one for PbPb run?
We don't have all changes in 80X or 76X.
Is the PR complete and good to go?

**Comment by deguio** [comment]
ciao @slava77 
this PR is essential. not for PbPb in particular, but in general for the L1T online monitoring.
this PR is supposed to be completed and already running at p5, but a last commit may come shortly.
I let @jingyucms comment about the ETA for that.

F.

**Comment by jingyucms** [comment]
ciao @slava77 
Yes this PR is essential for general L1T online monitoring, include HIon. Please merge it as possible as you can. There might be new update with explicit customization for explicit HIon tag. But the new one only gonna be couple lines, so please merge this one and I will make a new PR with much lighter code if needed.

**Comment by jingyucms** [comment]
@slava77 and the new one is not gonna affect anything offline

**Comment by slava77** [comment]
On 11/18/15 6:51 AM, Jingyu Zhang wrote:
> ciao @slava77 <https://github.com/slava77>
> Yes this PR is essential for general L1T online monitoring, include
> HIon. Please merge it as possible as you can. There might be new update
> with explicit customization for explicit HIon tag. But the new one only
> gonna be couple lines, so please merge this one and I will make a new PR
> with much lighter code if needed.
> 

This one seems late for pp reference run.
So, if a fix is needed for HI, it's needed before we can merge it.

**Comment by deguio** [comment]
ciao slava,
we prefer having this one merged. a later fix will come, but this one can already run and it is currently in use and plugged manually at p5.

**Comment by slava77** [comment]
On 11/18/15 7:57 AM, Federico De Guio wrote:
> ciao slava,
> we prefer having this one merged. a later fix will come, but this one
> can already run and it is currently in use and plugged manually at p5.
> 

If the changes are only for online and online already found a local
solution,
this is not a release driver anymore.
Maybe I'm missing something.
(comparisons didn't make it to judge if offline is affected).

**Comment by deguio** [comment]
well.. we have been plugging a lot of PRs manually and this is painful and not convenient for us.
this is not a release driver for offline, but in the end we use cmssw online as well.
F.

**Comment by jingyucms** [comment]
@deguio @slava77 update is coming, if you can give me half an hour

**Comment by deguio** [comment]
please test

**Comment by deguio** [comment]
+1

**Comment by jingyucms** [comment]
@slava77 @davidlange6 
Now the update is done. this one is for both pp and HI in 75x. If test pass, ready to be merged. @mulhearn can you also check and do "+1" again?

**Comment by slava77** [comment]
On 11/18/15 8:15 AM, Federico De Guio wrote:
> well.. we have been plugging a lot of PRs manually and this is painful
> and not convenient for us.
> this is not a release driver for offline, but in the end we use cmssw
> online as well.

Makes sense.
If there are no issues after the update, this will go to the next 75X.

**Comment by slava77** [comment]
@jingyucms 
please provide a more meaningful description of the PR (please edit the subject [use "Edit" button at the very top]) and add some meaningful description in the first comment block of the PR.
"backport 11756" is pretty useless because 11756 is closed

File: L1Trigger/Configuration/python/ValL1Emulator_cff.py [hunk 1]
```diff
@@ -122,11 +125,11 @@
...
-valGmtDigis.DTCandidates = cms.InputTag('gtDigis','DT')
-valGmtDigis.CSCCandidates = cms.InputTag('gtDigis','CSC')
-valGmtDigis.RPCbCandidates = cms.InputTag('gtDigis','RPCb')
-valGmtDigis.RPCfCandidates = cms.InputTag('gtDigis','RPCf')
-valGmtDigis.MipIsoData = 'gctDigis'
+valGmtDigis.DTCandidates = cms.InputTag('l1GtUnpack','DT')
+valGmtDigis.CSCCandidates = cms.InputTag('l1GtUnpack','CSC')
+valGmtDigis.RPCbCandidates = cms.InputTag('l1GtUnpack','RPCb')
+valGmtDigis.RPCfCandidates = cms.InputTag('l1GtUnpack','RPCf')
+valGmtDigis.MipIsoData = 'l1GtUnpack'
```
Comment (slava77) [inline L1Trigger/Configuration/python/ValL1Emulator_cff.py]: 80X version has caloStage1Digis here. Which one is correct?

Comment (jingyucms) [inline L1Trigger/Configuration/python/ValL1Emulator_cff.py]: @slava77 both are correct. currently we are using GCT (gctDigis) to read out calo regions(RCT). But in MC, we can do both gctDigis and caloStage1Digis. There is no difference.

File: DQM/L1TMonitor/src/L1THIonImp.cc [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+      int totp = 0;
+      int tot = 0;
+      if (it->gctEta() < 4){
+	totm += it->et();
+      }
+      if (it->gctEta() > 17){
+	totp += it->et();
+      }
+      if (it->gctEta() < 4 && it->gctEta() > 17){
+	tot += it->et();
```
Comment (slava77) [inline DQM/L1TMonitor/src/L1THIonImp.cc]: static analyzer points out that this is useless: indeed, you can just add totm and totp and drop this if and tot.

**Comment by jingyucms** [comment]
@slava77 dropped, please test again.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
```
error: unused variable 'tot' [-Werror=unused-variable]
       int tot = 0;

```
please fix

**Comment by slava77** [comment]
... please, at least compile before updating a PR

**Comment by slava77** [comment]
@cmsbuild please test
thanks for the quick fix.
It compiles now

File: DQM/Integration/python/clients/l1tstage1emulator_dqm_sourceclient-live_cfg.py [hunk 3]
```diff
@@ -219,8 +227,26 @@
...
+    process.load('Configuration/StandardSequences/FrontierConditions_GlobalTag_condDBv2_cff')
+    process.GlobalTag.toGet = cms.VPSet(
+        cms.PSet(
+            record  = cms.string("L1TCaloParamsRcd"),
+            tag     = cms.string("L1TCaloParamsStage1HI_CRAFT09_hlt"),
+            connect = cms.string("frontier://(proxyurl=http://localhost:3128)(serverurl=http://localhost:8000/FrontierOnProd)(serverurl=http://localhost:8000/FrontierOnProd)(retrieve-ziplevel=0)(failovertoserver=no)/CMS_CONDITIONS")
+            )
+        )
+else:
+    process.l1ExpertDataVsEmulatorStage1.remove (process.l1tHIonImp)
```
Comment (slava77) [inline DQM/Integration/python/clients/l1tstage1emulator_dqm_sourceclient-live_cfg.py]: I don't think this is working.
I get l1tHIonImp module running both in pp and in HeavyIonsRun2 configurations

Comment (slava77) [inline DQM/Integration/python/clients/l1tstage1emulator_dqm_sourceclient-live_cfg.py]: sorry, nevermind, this is an online config ; I was looking at offline DQM

File: DQM/L1TMonitor/python/L1TEmulatorMonitor_cff.py [hunk 4]
```diff
@@ -72,36 +84,40 @@
...
                                 )
 
 # for stage1
 l1ExpertDataVsEmulatorStage1 = cms.Sequence(
-    caloStage1Digis*
-    caloStage1LegacyFormatDigis*
+    #caloStage1Digis*
+    #caloStage1LegacyFormatDigis*
     l1TdeStage1Layer2 +
+    l1tHIonImp +
```
Comment (slava77) [inline DQM/L1TMonitor/python/L1TEmulatorMonitor_cff.py]: does it make sense to have it running for pp?

**Comment by slava77** [comment]
For both pp and HeavyIons (Run2) configurations I see the following additions/removal of running modules in the "reco+dqm" step

(with time/event included) using 251721 in this case as input processed with HeavyIonsRun2 prompt reco configuration
```
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
     removed      -0.02%         1.10 ms/ev ->         0.00 ms/ev l1compare
     removed      -0.00%         0.09 ms/ev ->         0.00 ms/ev l1ExtraDQM
     removed      -0.06%         2.84 ms/ev ->         0.00 ms/ev l1TdeRCTRun1
     removed      -0.00%         0.11 ms/ev ->         0.00 ms/ev l1tGct
     removed      -0.08%         3.79 ms/ev ->         0.00 ms/ev l1GtHwValidation
     removed      -0.07%         2.98 ms/ev ->         0.00 ms/ev valGtDigis
     removed      -0.02%         0.78 ms/ev ->         0.00 ms/ev valGctDigis
     removed      -0.01%         0.38 ms/ev ->         0.00 ms/ev l1demon
     removed      -0.01%         0.23 ms/ev ->         0.00 ms/ev dqmL1ExtraParticles
     removed      -0.00%         0.10 ms/ev ->         0.00 ms/ev l1tRctRun1
     removed      -0.00%         0.10 ms/ev ->         0.00 ms/ev l1TdeGCT
       added      +0.01%         0.00 ms/ev ->         0.32 ms/ev rctDigis
       added      +0.02%         0.00 ms/ev ->         0.67 ms/ev l1tPUM
       added      +0.01%         0.00 ms/ev ->         0.45 ms/ev l1demonstage1
       added      +0.01%         0.00 ms/ev ->         0.33 ms/ev l1tRctfromRCT
       added      +0.06%         0.00 ms/ev ->         2.80 ms/ev l1TdeRCTfromRCT
       added      +0.00%         0.00 ms/ev ->         0.18 ms/ev l1ExtraDQMStage1
       added      +0.08%         0.00 ms/ev ->         3.76 ms/ev l1Stage1GtHwValidation
       added      +0.00%         0.00 ms/ev ->         0.16 ms/ev valCaloStage1LegacyFormatDigis
       added      +0.00%         0.00 ms/ev ->         0.14 ms/ev l1tStage1Layer2
       added      +0.07%         0.00 ms/ev ->         2.94 ms/ev l1TdeRCT
       added      +0.01%         0.00 ms/ev ->         0.26 ms/ev simRctUpgradeFormatDigis
       added      +0.00%         0.00 ms/ev ->         0.18 ms/ev l1tHIonImp
       added      +0.00%         0.00 ms/ev ->         0.21 ms/ev l1TdeStage1Layer2
       added      +0.97%         0.00 ms/ev ->        43.15 ms/ev simCaloStage1Digis
       added      +0.01%         0.00 ms/ev ->         0.23 ms/ev l1GtUnpack
       added      +0.09%         0.00 ms/ev ->         3.81 ms/ev valStage1GtDigis
       added      +0.01%         0.00 ms/ev ->         0.26 ms/ev dqmL1ExtraParticlesStage1
       added      +0.00%         0.00 ms/ev ->         0.11 ms/ev l1tRct
       added      +0.03%         0.00 ms/ev ->         1.19 ms/ev l1compareforstage1

```

@jingyucms @mulhearn are all of these add/remove meaningful?

**Comment by slava77** [comment]
More oddities:
- DQM plots for run1 inputs change, showing fairly dead/bad L1T monitoring. I think that run1 L1 monitoring should not be broken.
the plot from wflow 1000.0 
wf1000_l1temu
- HI test run 261626 (processed with HeavyIonsRun2 setup and 75X_dataRun2_PromptHI_v1 GT) is not clearly better
wf261626_gcterr
wf261626_rct
- earlier run2 (2015B) looks worse: expected if inputs are from stage1, but probably are inadequate for actual trigger performance
wf134702_l1temu



Probably good things:
- run2 (2015C) L1T plots change quite a bit but most look somewhat OK, more meaningful. Here is one which looks pretty different from 134.802
wf134802_l1temu

**Comment by slava77** [comment]
... one more from 134.802 (2015C): a not obviously an improvement 
wf134802_tau_etm

**Comment by jingyucms** [comment]
@slava77 You are right, L1THIonImp is not need in the offline config. So I removed it. 
For those funny plots, what on top of my heads for now are:
1) I am not sure if running on run1 data is still useful
2) for run261626, the unpacker should unpack data from board 2301 to get better comparison, I didn't change board Id to 2300 is because in the future, all data should be unpacked from 2300. And the board ID is hard coded in EventFilter/L1TRawToDigi/src/implementationforstage1/CaloSetup.cc . 2301 is just for these days. If running with offline config, another possibility is the condition is for PP. But actually, we don't care about it. 
3) for 2015 run BC data, it is really difficult to track down the issue. Because you will have large change to running with wrong stage1 condition, many condition for those runs should be configured by hand. Especially, run 254790 is a bad run if my memory is correct.

**Comment by jingyucms** [comment]
@slava77 I just tested run261626 with board ID 2301, HIon agreement is good.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by deguio** [comment]
+1

**Comment by mulhearn** [comment]
+1

**Comment by slava77** [comment]
" 1) I am not sure if running on run1 data is still useful" 
this requirement is still present.
postLS1Customs are still there.
it looks like the changes applied in this PR are still driven by the configuration (not hadrcoded and impossible to change for run1 vs run2)

**Comment by jingyucms** [comment]
close. working on #12505

**Pull request closed without merging**